### DISCUSSION TITLE
Is it possible to disable output validation errors?

### DISCUSSION THREAD

**Discussion by ssteuteville** [RobinTail/express-zod-api#695] [Q&A] [answered]
I'd like output validation errors to result in a console.error on prod environments and a failed response on dev environments.

We're using nosql and every once in awhile we get a type wrong, i'd like for the endpoint to just return the data and not go into a complete failure state when that happens.

**Comment by RobinTail** [comment, marked as answer]
Hello @ssteuteville .

If you allow your endpoint to return wrong types, then the consumer of your API will receive not the ones you promised. Thus you'd just move the errors from one place to another — your Frontend will have to deal with `null` instead of the expected `undefined`, for example.
Instead, I'd suggest to create more loose output types using `.or()`.
The types validation is the essential goal of the library. It's intended to operate in strict mode.

**Reply by elmeister** [reply]
Oh, it's actually a big deal for us too. We're slowly moving to this beautiful library, but if for some reason we will make an error in output - as of now it will work fine, since our clients probably already handle these cases, but if the whole payload will fail, it will introduce a new big issue.
The `.or()` thing is not an option as well, since we can't predict where exactly things will go wrong.
Would be nice to have option to still proceed with response,

**Reply by ssteuteville** [reply]
@RobinTail 

You didn't answer my question. My question is, is it possible to disable the errors? I'd like to get warnings that I can monitor instead. We have manual and automated QA processes in place, so typically when we miss something, it's already been verified that our clients are still working.

You responded to me with the express-zod-api opinion that is already obvious through code/functionality. I'd like to know if it's possible to override that opinion through a custom request handler or other customization setting.

**Reply by RobinTail** [reply]
It's not possible, @ssteuteville .
In my response I tried to explain to you why.

**Reply by michaelbarone** [reply]
This is a use case for my team as well.  we often have different developers working on the api layer and the front end applications.  Some endpoints can have an intended change in functionality/response or change with a bugfix, and if something is problematic we want to catch in dev with broken functionality, but we dont lose complete functionality if this change makes it into production and was not caught by QA.

allowing this as an option may allow additional use cases for your library and increase the use/adoption.

**Reply by ssteuteville** [reply]
> It's not possible, @ssteuteville . In my response I tried to explain to you why.

Ok thank you, sadly, I think this opinion is a little too strict for a lot of production systems. Without the ability to get warnings in production and errors in dev I'm likely going to have to move on to a different solution or use `output: z.object({}).passthrough()` for everything.

I really do love everything else about this library, but I can't have a prod API failing because we returned null instead of undefined by accident..

**Reply by RobinTail** [reply]
@ssteuteville ,

It is impossible due to ZOD library architecture to do what you want. Here is [the line](https://github.com/RobinTail/express-zod-api/blob/d3300a26b0d653a5fe3311a419956a1a2fa82015/src/endpoint.ts#L218) in my code:

```ts
this.outputSchema.parseAsync(output);
```

Output schema MAY contain transformations, therefore it requires the call of a parsing method on it.
However, even `.safeParseAsync()` wouldn't run transformations (ZodEffect) and ignore the validation issues at the same time.

> This method returns an object containing either the successfully parsed data or a ZodError instance containing detailed information about the validation problems.

There is no possibility for a partial validation, or turning errors into warnings or anything like that.

**Reply by gmazoni** [reply]
I hit the same question today to help me migrating a legacy API. Maybe allow a raw output of the handler in the result handler? So I can verify if the name of the error is OutputValidationError and pass a non error status code and the raw output of the endpoint handler in the final response? Like:

![image](https://github.com/RobinTail/express-zod-api/assets/40010791/2febdbad-9195-4764-9699-550118dbf950)

My workaroud was to make a HOF to use NODE_ENV to return `z.object({}).passthrough()` or a desired schema to the output.
![image](https://github.com/RobinTail/express-zod-api/assets/40010791/a0a6aad4-6912-4554-a10c-d34d5f39e365)
![image](https://github.com/RobinTail/express-zod-api/assets/40010791/3f90f797-88dc-463d-aa88-4134f6a797bd)

Losing `coerce` and other things, because it dont knows if is valid or not, so it must assume it's not always.

**Reply by RobinTail** [reply]
The correct way for checking the kind of error:

```ts
error instanceof OutputValidationError
```

Output validation is essential feature that ensures that the response comply the agreed contract.
However, the contract can be loose enough to allow uncertainty during the migration of legacy API,

@gmazoni

**Reply by gmazoni** [reply]
You are right, but it's worth nothing because the output is undefined in the ResultHandler when it occurs, thinking of getting the raw output in the result handler, we can not only ignore and force an output, but we can have the possibility to implement more logic, like, logging a warn, doing a fallback strategy, etc. Have you considered this? It can be a strategic resource to rollout the refactored endpoints, I can leave the desired contract and monitor/trace the cases where it's not agreed, then flip the switch to enforce. Imagining a scenario of making the contracts weaker an the beggining to avoid errors can leave lots of techinical debt in terms os schemas. After that I can fix the front-end that already deals with the wrong output of the backend.

**Reply by RobinTail** [reply]
The purpose of this library is strict compliance of data with the contract and consistency of responses.

> Have you considered this?

Yes, and here is my final resolution to this problem:

1. If it's not clear what kind of data the endpoint responds with, it is the issue of that API implementation, not the one of framework/library.
2. If it's required purposely to break some constraints in the framework/library for the time of figuring out what is wrong with your implementation, there is a way for doing this — forking the library, but not making it breakable in production for everyone.

![image](https://github.com/RobinTail/express-zod-api/assets/13189514/a1f1bff5-8abd-425a-9037-b0030dd8e46d)
